### PR TITLE
Support extracting the canonical URI from a document

### PR DESCRIPTION
**Pull request by Toflar**
I think once we start working on e.g. a new search implementation, this might come in handy.

/cc @ausi

### CODE REVIEW

File: core-bundle/src/Search/Document.php [hunk 1]
```diff
@@ -90,6 +90,30 @@ public function getContentCrawler(): Crawler
         return $this->crawler;
     }
 
+    public function extractCanonicalUri(): ?UriInterface
+    {
+        foreach ($this->getHeaders() as $key => $values) {
+            if ('link' === $key) {
+                foreach ($values as $value) {
+                    if (preg_match('@<(https?://(.+))>; rel="canonical"@', $value, $matches)) {
```
Comment (ausi) [inline core-bundle/src/Search/Document.php]:
```suggestion
                    if (preg_match('@<(https?://(.+))>;\s*rel="canonical"@', $value, $matches)) {
```

as far as I can see from the spec, the space before `rel=` seems to be optional? Not sure if spaces between around `=` are allowed too? `rel = "canonical"`

**Review by ausi — COMMENTED** [review]
`rel="canonical"` links don’t have to be absolute, so if we want a robust implementation we should probably resolve relative URLs to absolute ones?

**Comment by Toflar** [comment]
> `rel="canonical"` links don’t have to be absolute, so if we want a robust implementation we should probably resolve relative URLs to absolute ones?

According to Google, they have to be: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls#rel-canonical-link-method

I think that's good practice. And if you don't, then it's just not gonna work - no big deal.

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Pull request merged**